### PR TITLE
Make an entry in /etc/group when we modify /etc/passwd

### LINKED ISSUE #7389: userns=keep-id creates unnamed user group
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

Container created with option `--userns=keep-id` gets the current user with the correct UID and GUID but the group does not have the user's name.

More info from the Toolbox realm: https://github.com/containers/toolbox/issues/523

**Steps to reproduce the issue:**

1. `podman run --rm --userns=keep-id alpine id`

**Describe the results you received:**

```
uid=1000(omichal) gid=1000
```

**Describe the results you expected:**

```
uid=1000(omichal) gid=1000(omichal)
```

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
Version:      2.1.0-dev
API Version:  1
Go Version:   go1.15rc2
Built:        Thu Jan  1 01:00:00 1970
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.16.0-dev
  cgroupVersion: v2
  conmon:
    package: conmon-2.0.21-0.3.dev.git5a6b2ac.fc33.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.21-dev, commit: 5c03a23398b94cf869f071128631ef7fd9153b3b'
  cpus: 8
  distribution:
    distribution: fedora
    version: "33"
  eventLogger: file
  hostname: harry-work
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 5.8.0-1.fc33.x86_64
  linkmode: dynamic
  memFree: 4838907904
  memTotal: 16656273408
  ociRuntime:
    name: crun
    package: crun-0.14.1-2.fc33.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 0.14.1
      commit: 598ea5e192ca12d4f6378217d3ab1415efeddefa
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  remoteSocket:
    exists: true
    path: /run/user/1000/podman/podman.sock
  rootless: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.4-4.dev.giteecccdb.fc33.x86_64
    version: |-
      slirp4netns version 1.1.4+dev
      commit: eecccdb96f587b11d7764556ffacfeaffe4b6e11
      libslirp: 4.3.1
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.0
  swapFree: 12363067392
  swapTotal: 12708732928
  uptime: 46h 1m 14.24s (Approximately 1.92 days)
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - registry.centos.org
  - docker.io
store:
  configFile: /var/home/omichal/.config/containers/storage.conf
  containerStore:
    number: 13
    paused: 0
    running: 2
    stopped: 11
  graphDriverName: overlay
  graphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: fuse-overlayfs-1.1.0-11.dev.git800011b.fc33.x86_64
      Version: |-
        fusermount3 version: 3.9.3
        fuse-overlayfs: version 1.1.0
        FUSE library version 3.9.3
        using FUSE kernel interface version 7.31
  graphRoot: /var/home/omichal/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 11
  runRoot: /run/user/1000/containers
  volumePath: /var/home/omichal/.local/share/containers/storage/volumes
version:
  APIVersion: 1
  Built: 0
  BuiltTime: Thu Jan  1 01:00:00 1970
  GitCommit: ""
  GoVersion: go1.15rc2
  OsArch: linux/amd64
  Version: 2.1.0-dev
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-2.1.0-0.169.dev.git162625f.fc33.x86_64
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Fedora Silverblue Rawhide (Rawhide.20200811.n.0)

### LINKED ISSUE #7503: User from "generateCurrentUserPasswdEntry" created with duplicate UID
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

Since 2.0.5 podman creates a user account when `--userns=keep-id` is set, with the UID of the user on the host system. The algorithm doesn't check if there's _already_ a user with the same UID inside the container. As a result in the end there might be two user accounts with the same UID.

**Steps to reproduce the issue:**

1. Given your UID is 1000, run a container off "quay.io/podman/stable" (which already has a user with UID 1000)
```console
$ id -u
1000
$ podman run --rm -it --user=root --userns=keep-id quay.io/podman/stable tail -n2 /etc/passwd
podman:x:1000:1000::/home/podman:/bin/bash
USER:x:1000:995:USER:/:/bin/sh
```

**Describe the results you received:**
The `/etc/passwd` in the containers now has two users with UID 1000.

**Describe the results you expected:**
Not sure how this should be handled, I guess either:
- A: Not create a new entry, if there's already an entry with this UID
- B: Abort the creation of the container with an error.

**Additional information you deem important (e.g. issue happens only occasionally):**
I've noticed this behaviour when investigating https://github.com/containers/podman/issues/7499.
I'm not sure if the duplicate UIDs are a problem, and how severe. Maybe it's actually ok and this ticket can be closed. Just wanted to bring it to your attention.

This functionality was added in https://github.com/containers/podman/commit/3dfd8630a51a37734ad8c51162c4d004b8ffffb2 cc @rhatdan @vrothberg 

**Output of `podman version`:**
```
Version:      2.0.5
API Version:  1
Go Version:   go1.15
Git Commit:   776abc52106ec7652ced6dbc0869020123ed393d
Built:        Tue Aug 25 20:58:48 2020
OS/Arch:      linux/amd64

```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.15.1
  cgroupVersion: v1
  conmon:
    package: Unknown
    path: /usr/bin/conmon
    version: 'conmon version 2.0.20, commit: 13244db638cf987c415298a3c23393ae5abeb885'
  cpus: 12
  distribution:
    distribution: arch
    version: unknown
  eventLogger: file
  hostname: njamdesk
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 995
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 5.8.5-arch1-1
  linkmode: dynamic
  memFree: 5362950144
  memTotal: 33520361472
  ociRuntime:
    name: runc
    package: Unknown
    path: /usr/bin/runc
    version: |-
      runc version 1.0.0-rc92
      commit: ff819c7e9184c13b7c2607fe6c30ae19403a7aff
      spec: 1.0.2-dev
  os: linux
  remoteSocket:
    path: /run/user/1000/podman/podman.sock
  rootless: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: Unknown
    version: |-
      slirp4netns version 1.1.4
      commit: b66ffa8e262507e37fca689822d23430f3357fe8
      libslirp: 4.3.1
      SLIRP_CONFIG_VERSION_MAX: 3
  swapFree: 0
  swapTotal: 0
  uptime: 5h 45m 58.77s (Approximately 0.21 days)
registries:
  search:
  - docker.io
  - registry.fedoraproject.org
  - quay.io
  - registry.access.redhat.com
  - registry.centos.org
store:
  configFile: /home/reto/.config/containers/storage.conf
  containerStore:
    number: 29
    paused: 0
    running: 3
    stopped: 26
  graphDriverName: overlay
  graphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: Unknown
      Version: |-
        fusermount3 version: 3.9.3
        fuse-overlayfs: version 1.1.0
        FUSE library version 3.9.3
        using FUSE kernel interface version 7.31
  graphRoot: /home/reto/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: btrfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 95
  runRoot: /run/user/1000/containers
  volumePath: /home/reto/.local/share/containers/storage/volumes
version:
  APIVersion: 1
  Built: 1598381928
  BuiltTime: Tue Aug 25 20:58:48 2020
  GitCommit: 776abc52106ec7652ced6dbc0869020123ed393d
  GoVersion: go1.15
  OsArch: linux/amd64
  Version: 2.0.5
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**
```
$ pacman -Q podman       
podman 2.0.5-1
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**

physical

### PR DESCRIPTION
**Pull request by mheon**
To ensure that the user running in the container ahs a valid entry in /etc/passwd so lookup functions for the current user will not error, Podman previously began adding entries to the passwd file. We did not, however, add entries to the group file, and this created problems - our passwd entries included the group the user is in, but said group might not exist. The solution is to mirror our logic for /etc/passwd modifications to also edit /etc/group in the container.

Unfortunately, this is not a catch-all solution. Our logic here is only advanced enough to *add* to the group file - so if the group already exists but we add a user not a part of it, we will not modify that existing entry, and things remain inconsistent. We can look into adding this later if we absolutely need to, but it would involve adding significant complexity to this already massively complicated function.

While we're here, address an edge case where Podman could add a user or group whose UID overlapped with an existing user or group.

Fixes #7503
Fixes #7389

### CODE REVIEW

**Comment by mheon** [comment]
Still needs tests

**Comment by rhatdan** [comment]
We need e2e and/or system tests for this as well, to verify that we handle the situations correctly.

Group Name added to container (--keep-id and --user=1234)
--read-only
Group Name exists in container
/etc/group does not exist in container

**Comment by mheon** [comment]
Tests added, and converted to fix #7499 as well (generate a `*` instead of `x` for passwd entries).

File: libpod/container_internal_linux.go [hunk 1]
```diff
@@ -1499,23 +1506,171 @@ func (c *Container) getHosts() string {
...
+	if err != nil {
+		return "", 0, nil
+	}
+
+	if addedGID != 0 && addedGID == int(gid) {
+		return "", 0, nil
+	}
+
+	// Check if the group already exists
+	_, err = lookup.GetGroup(c.state.Mountpoint, group)
```
Comment (rhatdan) [inline libpod/container_internal_linux.go:1608]: Nit can you do this on one line.
if _, err := lookup.GetGroup(c.state.Mountpoint, group); err !=nil {
...
}

File: libpod/container_internal_linux_test.go [hunk 2]
```diff
@@ -29,16 +29,42 @@ func TestGenerateUserPasswdEntry(t *testing.T) {
...
+		},
+	}
+	group, _, err := c.generateUserGroupEntry(0)
+	if err != nil {
+		t.Fatal(err)
+	}
+	assert.Equal(t, group, "456:x:456:123\n")
+
+	c.config.User = "567"
+	user, _, _, err = c.generateUserPasswdEntry(0)
```
Comment (TomSweeneyRedHat) [inline libpod/container_internal_linux_test.go:65]: lint isn't happy, I think you need the below:

```suggestion
	user, _, _, err1 := c.generateUserPasswdEntry(0)
	if err1 != nil {
```

**Comment by mheon** [comment]
This is starting to go green.

**Comment by rhatdan** [comment]
LGTM
@ashley-cui @TomSweeneyRedHat @baude @vrothberg @jwhonce @QiWang19 PTAL

**Comment by TomSweeneyRedHat** [comment]
LGTM

File: libpod/container_internal_linux.go (hunk 1 above)
Comment (QiWang19) [reply libpod/container_internal_linux.go:1608]: nits still there

**Comment by QiWang19** [comment]
LGTM
